### ISSUE TITLE
Branch length distribution between parent and child nodes within the same timeslice

### ISSUE THREAD

**Issue by hyanwong** [tskit-dev/tsdate#241]
When a parent and child node are in different timeslices (say $a$ and $b$), I think we simply assume that the branch length between them is the distance between the midpoints of the time slices (or something similar). This is an approximation, of course: what we really want is the convolution of the smoothed pdf of the parent within timeslice $a$ with the smoothed pdf of the child within timeslice $b$. We would then integrate across the convolution to get the likelihood of the observed number of mutations on that edge. I presume that the midpoint method is fine in this case. However, we never properly sorted out how to deal with the likelihoods when a parent and a child are in the *same* timeslice (end hence the distance between midpoints is zero).

This may come to a head when we start to allow historical samples, which can be present halfway up a timeslice. In this case, we have previously (in a non-merged PR) decided simply to truncate the prior of the parent of the historical sample, so that parental timeslices younger than the child have a probability of zero. We do nevertheless want to allow a small probability that the parent could be in the same timeslice (this probability would obviously be smaller the nearer the child is to the top of the timeslice). I think we can do this using some form of quadrature, where we fit a curve to the parent's prior distribution, and cut off that curve at the time of the oldest historical child. Considering the distribution of branch lengths between the child and the parent should be easy: we can integrate over the possible branch lengths weighted by their probability (and we can probably treat the probability as a constant).

The other case to think about is where we have a child and a parent neither of which are fixed, within the same timeslice. In this case we should only consider the convolution of parent and child times when the parent is strictly older than the child. This feels like it would be a weird distribution. @nspope: does this make sense, and have you ever through about this latter problem?

If we can solve this, I think it is the only remaining issue to fix the timeslice interpretation problem. I hope this might fix some of the weird results we get for very young node times, when I fear we overestimate node times somewhat.

**Comment by hyanwong** [comment]
We try to fits this currently using an "epsilon" value, but I think this is the wrong thing to do. Here's the relevant bit of code: https://github.com/tskit-dev/tsdate/blob/494f010933bc5a974c1a429ed69e7fa7db50f4f6/tsdate/core.py#L83

**Comment by nspope** [comment]
To make sure I understand, the question is what is the likelihood after integrating parent, child ages over the time slice? Given that parent, child ages are iid uniform in that time slice? 

That is, if $t$ is the age of the parent, $v_1 \dots v_n$ are the ages of the $n$ children, $y_1 \dots y_n$ are counts of mutations on the edges leading to each child; $[a, b)$ is the timeslice; $\mu_i$ is the mutation rate weighted by edge span (e.g. sequence length); then

$$ \int_a^b \frac{1}{b-a} \left( \prod_{i=1}^n \int_a^t \frac{1}{b-a} p(y_i \mid \mu_i (t - v_i)) dv_i \right) dt $$

where $p(. \mid \lambda)$ is a Poisson process with intensity $\lambda$?

**Comment by nspope** [comment]
To rephrase/think out loud, my understanding is that the "correct" (e.g. continuous-space) posterior for the age of node $j$ in the inside pass would have the form, $$p(t_j \mid y_{/j}) \propto p(t_j) \prod_{i \in C_j} \int_0^{t_j} p(t_i \mid y_{/i}) \ell(y_i \mid t_j - t_i) dt_i $$ where $C_j$ is the set of children of $j$-th node, $y_i$ is the mutation count on the edge subtending $i$-th node, $y_{/j}$ is the set of mutation counts for edges in all subtrees rooted at the $j$-th node. So that $p(t_j)$ is the prior, $p(t_i \mid y_{/i})$ is the posterior conditioned on mutations in the subtrees rooted at the $i$-th node, and $\ell(y \mid .)$ is the Poisson PMF adjusted for mutation rate and edge span.

But, the integrals are convolutions which don't have a closed form. So, this is approximated by:
1. Get a grid of timepoints (e.g. from quantiles of the per-node priors)
2. Calculate all positive time differences along this grid
3. Add a small positive amount $\epsilon$ to each time difference, so that the minimum time difference is positive
4. Use these time differences as the set of possible branch lengths for a Poisson PMF
5. Do the convolutions along this discretization -- which is a set of atoms (discrete timepoints) rather than intervals -- somehow weighting by magnitude of time difference (?)

This makes sense to me, but there's always going to be some bias introduced by choice of discretization and $\epsilon$. One alternative is to calculate the posterior over intervals rather than discrete atoms, which I think is what @hyanwong is suggesting above. That is, use an approximate posterior $\tilde{p}(t_j \mid y_{/i})$ that is piecewise uniform, e.g. $$\tilde{p}(t_i \mid y_{/i}) = \sum_{k=1}^K \mathbb{I}[a_k \leq t_i < a_{k+1}] \frac{P_{ik}}{a_{k+1} - a_{k}}$$ where $P_{ik}$ are the posterior weights stored in the grid for node $i$ across the $K$ finite breakpoints $a_1 \dots a_K$. These would be calculated by convolution as before, except the likelihood would not be evaluated pointwise but instead integrated over the piecewise-uniform distributions:
$$\tilde{p}(t_j \mid y_{/j}) \propto p(t_j) \prod_{i \in C_j} \int_0^{t_j} \tilde{p}(t_i \mid y_{/i}) \ell(y_i \mid t_j - t_i, \mu_{i}) dt_i = p(t_j) \prod_{i \in C_j}  \sum_{k=1}^K \frac{P_{ik} L_{ik}(t_j)}{a_{k+1}-a_k}$$
where $L_{ik}(t_j) = \int_{a_k}^{a_{k+1}} \mathbb{I}[t_i < t_j] \ell(y_i \mid t_j - t_i) dt_i$ is the integrated likelihood conditional on $t_j$ (I think this has a closed form, or at least can be written in terms of easy-to-compute special functions). Then, integrating over the parent age $j$ in time intervals gives posterior weights for the parent that are $P_{jk} = \int_{a_k}^{a_{k+1}} \tilde{p}(t_j \mid y_{/j}) dt_j$. This could be done with quadrature and it doesn't require discretizing the prior or setting $\epsilon$. However, this strategy would be a lot more expensive on a per-node basis, because it couldn't use pre-computed likelihoods.

**Comment by hyanwong** [comment]
That's a great summary @nspope. You are absolutely right that using the time differences allows us to pre-compute the likelihoods, and I had forgotten that's why we use a time *points* array rather than using time slices. The one thing you didn't mention was that the time points approach therefore has a problem with the oldest time point, whereas the timeslice approach can simply have a timeslice that goes to infinity.

The epsilon hack has (I think) problems with inaccuracy when nodes are in the same timeslice, so I have previously worries that we are forcing parents and their children to be more likely to be in different timeslices. I don't know if we ever quite fixed this.

**Comment by hyanwong** [comment]
Note that with the current timepoints approach, I'm not 100% sure we have the exactly correct expressions for the conversion of likelihoods-at-timepoints to probabilities or for the calculation of the mean / variance from the probabilities, as we are treating both using a timeslice approach (and normalising so that the sum of probabilities over all timeslices is one, and the mean does not include an estimate of the weight of the topmost time slice.

**Comment by nspope** [comment]
I think with the integrated intervals you can still do some precalculation (for all intervals prior to a point $t$, one can calculate & reuse the integrated likelihood) which should narrow the gap somewhat. There will always be some calculation that needs to happen on the fly, though.

With regards to the last timepoint, one advantage of *not* discretizing the prior (which is possible if we work with intervals) is that it makes the convolution integrals over that last time slice convergent. Otherwise, I think they'd diverge when there's 0 mutations on a branch.

It occurs to me that this is basically a finite-element method -- it's building a nonparametric variational approximation to the solution using piecewise-constant elements. It'd be a bit of work to actually turn this into something that scales, but should be easy enough to get a (slow) working example going. Then we could see if this strategy helps with some of the toy problems where there's obvious bias.

**Comment by hyanwong** [comment]
We haven't tried using the (weighted?) midpoints of the timeslices as the discretisation points, with some sort of reasonable exponential weighting for the last timeslice, and a more sophisticated approach for parent+child at the same timepoint.

We also haven't properly accounted for cases where a historical sample is present between discretised timepoints. Given there aren't usually many historical samples, this could probably be done without using the pre-computed likelihoods.

**Comment by hyanwong** [comment]
p.s.: I agree that it might be useful to see if a slow working example addresses some of the problems (and we should have test cases that identify the problems, I think). I'm not sure I have the capacity to develop a slow working example at the moment, though.

**Comment by nspope** [comment]
Agreed, there are easier approaches to try first! I do think it'd be good to be explicit about the support of the approximation, though -- whether it's for intervals, atoms, or some combination thereof.

**Comment by hyanwong** [comment]
> I do think it'd be good to be explicit about the support of the approximation, though -- whether it's for intervals, atoms, or some combination thereof.

Yes, I agree, and I don't think this is done fully yet. I also note that the likelihoods aren't precalculated above sample ("fixed") nodes (see [here](https://github.com/tskit-dev/tsdate/blob/b9bf7d83f55f940a90565be8802e96963a6f9871/tsdate/core.py#L178)). So I do start to wonder if we should use the *midpoints* of timeslices (or maybe midpoints on a log scale?) for the timepoints in the likelihood calculations. Then we could pick a reasonable point to represent the last timeslice. E.g. we could calculate a timeslice array like this:

```
[0, 10, 20, 100]
```

With an associated *probabilities* of being in the timeslices like:

```
0-10: p = 0.5
10-20: p=0.25
20-100: p=0.2,
100-infinity: p=0.05
```

Then we set a *midpoints* array for use in calculating the likelihoods given the poisson probabilities of M mutations on an edge:

```
midpoints: [5, 15, 60, ???]  # probably better to take the midpoint on a log scale
```

We then shouldn't need an epsilon for the sample nodes at time 0 (and since we aren't precalculating it, we could use a more exact method anyway). We would need to think hard about how to deal with the probability of a non-fixed node being in the same timeslice as its parent. But again we could work out something based on e.g. a uniform distribution (or perhaps uniform in log space) of possible child /parent times within the timeslice, and we could cache that.

**Comment by nspope** [comment]
Yes, I think this is a good idea. If that's the case, then the prior grid stores the timeslice breaks, whereas the likelihood class stores the midpoints of time slices? I definitely lean towards making the midpoints in linear space, not logspace, as it'd make the connection to a piecewise-uniform distribution much cleaner. I'll work out what some of the integrated likelihood stuff would look like, should have an idea soon of whether it could be used for the case when parent and child are in the same timeslice.

**Comment by hyanwong** [comment]
> If that's the case, then the prior grid stores the timeslice breaks, whereas the likelihood class stores the midpoints of time slices? 

I guess so, yes.

> I definitely lean towards making the midpoints in linear space, not logspace, as it'd make the connection to a piecewise-uniform distribution much cleaner. 

I tend to think that log time is better for all this coalescent stuff, and reflects the variability in estimates much better. If *everythign* is done in log time, including the piecewise-uniform distribution, then I think this will work out OK? But I don't mind going for linear, especially as a first pass.

> I'll work out what some of the integrated likelihood stuff would look like, should have an idea soon of whether it could be used for the case when parent and child are in the same timeslice.

That would be fantastic, thanks. This has been on my medium-priority to-do list for some time. It would be great if you had time to think about it.

**Comment by nspope** [comment]
>If everythign is done in log time, including the piecewise-uniform distribution, then I think this will work out OK?

If you take a distribution that's uniform in log-space and transform it back it linear space, it's no longer uniform because of change of variables:
$$f(u) = \frac{\mathbb{I}[a \leq u < b]}{(a - b)},~~t = \exp{u} \implies f(t) = \frac{1}{t} \frac{\mathbb{I}(\exp{a} \leq t < \exp{b}]}{\exp{b} - \exp{a}} $$
what this'd look like (in linear space) is a piecewise distribution where the greatest mass in each piece is at the start and decays over the timeslice, making a sawtooth-like pattern. Because the timepoints are more-or-less arbitrary, this doesn't make much sense to me.

It does makes sense to choose the timepoints with logarithmic spacing, but have each piece be uniform in linear space -- however, I think the current strategy of choosing timepoints from quantiles from the prior is nicer.

**Comment by hyanwong** [comment]
I realise that if you back-transform the distribution it will indeed have different probability masses over time. I'm just saying that it makes sense to switch entirely to log space and think about all the patterns in this way. Or, to put it another way, if you log-transform the piecewise uniform distribution, you will end up with a sawtooth in the other direction. Both are features of the discretisation. I'm usually happier to simply think of time-into-the-past on a log scale in all contexts, and never even worry about what it looks like in linear time.

But, as I said, it's a pretty minor point. I'm happy to go with linear time, and come up with an exception for the oldest timeslice.

NB - another reason to think in log time: if the timelines are very wide (e.g. there's only one or two) then when calculating the mean, there is much more variability in the estimates that contribute to the top end of the timeslice than the more recent end.

**Comment by nspope** [comment]
>Or, to put it another way, if you log-transform the piecewise uniform distribution, you will end up with a sawtooth in the other direction. fOr someone entirely thinking in log space, this might seem equally odd.

Fair enough. If the approximation is defined in logspace, then the posterior mean should be calculated in logspace, too. That doesn't really matter if the expectation is over a bunch of discrete time points, but if it's over time slices then the choice of measure will have a (possibly large) impact. I don't mean to be pedantic over a possibly minor issue -- just think it'd be good to hammer out what exactly we're going for here, as I'm sure this topic will come up again!

**Comment by hyanwong** [comment]
> Not to be pedantic over a possibly minor issue -- just want to hammer out what exactly we're going for here, as I'm sure this topic will come up again!

Yes, it's as well to tighten up all these definitions.

Note that the prior is (by default, and empirically best) a lognormal, which sort-of implies that we should be thinking in log time by default. Log time does, however, cause problems for samples at time 0.

**Comment by nspope** [comment]
I implemented a recycled quadrature scheme in [this branch](https://github.com/nspope/tsdate/tree/recycled-quadrature).  The idea is taken from this [paper](https://academic.oup.com/gbe/article/8/5/1338/2939555), where they're calculating an integral of a latent trait up a tree. The approach is to use quadrature weights (from Simpson's rule) for each convolution, rather than the simple summation over timepoints. With a fixed quadrature grid, the estimates can be reused at each iteration of the dynamic program, so it's very similar to the current algorithm. Ideally this would do a better job at interpolating in parts of the grid where time is sparsely sampled, like the tail and very recent times.

Unfortunately, it doesn't really seem to improve things -- when the mutation rate is realistic (1e-8 or so) the estimates are a bit better in the tails, but are biased upwards in the center of the time distribution.  When mutation rate is high, both approaches converge to the truth. Maybe doing quadrature over log-time would work a bit better-- I'll mess around with the idea some more.

**Comment by hyanwong** [comment]
Thanks @nspope . Have you tried seeing if it improves things with a very sparse time grid? Also thanks for experimenting with this. It's very useful. If we think it's the right way to go, and is technically more accurate, then we should think about merging it anyway. It could be that there are other things, such as the "loopy belief propagation" issue in #165, that are affecting accuracy and confounding the results. For this reason it is probably worth trying on a large single tree, where #165 should not be relevant.

**Comment by nspope** [comment]
Great point @hyanwong, quadrature does indeed seem to do better for single trees. I guess the discrepancy versus tree sequences might have to do with the fact that the children/parents of a given node in an ARG can subtend overlapping sets of mutations; which of course can't happen in a single tree. I jotted down some [notes](https://github.com/tskit-dev/tsdate/files/10483391/quadrature.pdf) in an attempt to make this a bit more precise.

One possible advantage to using quadrature (assuming we can work this out) is that the "density" at a particular time point is approximated, rather than the probabilities at a finite set of possible timepoints. What this means is that a separate time grid could be used for each node; and the "density" for each node could be interpolated to the grids of parent/child nodes at each step of the dynamic program. Because likelihoods are precomputed by edge, I don't think this would reduce efficiency too much. In fact, it might improve it, because a smaller set of a timepoints could be used for each node. If the interpolation doesn't  distort the approximations too much, this would effectively solve the "bunching up" that happens at very recent times.
